eo-phi-normalizer/app/Main.hs

```diff
@@ -1,111 +1,222 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
 {-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE ImplicitParams #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE OverloadedLists #-}
 {-# LANGUAGE OverloadedRecordDot #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
 {-# LANGUAGE RecordWildCards #-}
+{-# OPTIONS_GHC -Wno-partial-fields #-}
 {-# OPTIONS_GHC -Wno-type-defaults #-}
 
-module Main where
+module Main (main) where
 
 import Control.Monad (unless, when)
 import Data.Foldable (forM_)
 
 import Control.Lens ((^.))
 import Data.Aeson (ToJSON)
-import Data.Aeson.Text (encodeToLazyText)
+import Data.Aeson.Encode.Pretty (Config (..), Indent (..), defConfig, encodePrettyToTextBuilder')
 import Data.List (nub)
+import Data.String.Interpolate (i)
+import Data.Text.Internal.Builder (toLazyText)
 import Data.Text.Lazy.Lens
-import Language.EO.Phi (Object (Formation), Program (Program), defaultMain, parseProgram, printTree)
-import Language.EO.Phi.Rules.Common (Context (..), applyRules, applyRulesChain)
+import GHC.Generics (Generic)
+import Language.EO.Phi (Object (Formation), Program (Program), parseProgram, printTree)
+import Language.EO.Phi.Metrics.Collect (collectMetrics)
+import Language.EO.Phi.Rules.Common (applyRules, applyRulesChain)
+import Language.EO.Phi.Rules.Common qualified as Common
 import Language.EO.Phi.Rules.Yaml (RuleSet (rules, title), convertRule, parseRuleSetFromFile)
-import Options.Generic
-import System.IO (IOMode (WriteMode), hClose, hPutStr, hPutStrLn, openFile, stdout)
+import Options.Applicative
+import Options.Applicative.Types qualified as Optparse (Context (..))
+import System.IO (IOMode (WriteMode), getContents', hFlush, hPutStr, hPutStrLn, openFile, stdout)
 
-data CLINamedParams = CLINamedParams
+data CLI'TransformPhi = CLI'TransformPhi
   { chain :: Bool
-  , rulesYaml :: Maybe String
-  , outPath :: Maybe String
+  , rulesPath :: String
+  , outputFile :: Maybe String
   , single :: Bool
   , json :: Bool
+  , inputFile :: Maybe FilePath
```

Comment:
Why did you make it a named parameter instead of positional?

Reply:
I want to support reading a program:

- from a file: `--input-file program.phi`
- from an option: `--program "{ ⟦ ⟧ }"`
- from stdin: `--input-file -`

Reply:
I don't agree with these. I think what we already had is more idiomatic and we already supported all three cases:
1. `normalizer transform program.phi`
2. `echo "{ ⟦ ⟧ }" | normalizer transform`
3. `normalizer transform`

This is what I think is the approach followed by most programs I used, and what we already supported